nixos/modules/services/system/userborn.nix

```diff
@@ -0,0 +1,156 @@
+{ utils, config, lib, pkgs, ... }:
+
+let
+
+  cfg = config.services.userborn;
+  userCfg = config.users;
+
+  userbornConfig = {
+    groups = lib.mapAttrsToList
+      (username: opts: {
+        inherit (opts)
+          name
+          gid
+          members
+          ;
+      })
+      config.users.groups;
+
+    users = lib.mapAttrsToList
+      (username: opts: {
+        inherit (opts)
+          name
+          uid
+          group
+          description
+          home
+          password
+          hashedPassword
+          hashedPasswordFile
+          initialPassword
+          initialHashedPassword
+          ;
+        isNormal = opts.isNormalUser;
+        shell = utils.toShellPath opts.shell;
+      })
+      config.users.users;
+
+  };
+
+  userbornConfigJson = pkgs.writeText "userborn.json" (builtins.toJSON userbornConfig);
+
+  immutableEtc = config.system.etc.overlay.enable && !config.system.etc.overlay.mutable;
+  passwordFilesLocation = if immutableEtc then cfg.immutablePasswordFilesLocation else "/etc";
+  # The filenames created by userborn.
+  passwordFiles = [ "group" "passwd" "shadow" ];
+
+in
+{
+
+  options.services.userborn = {
+
+    enable = lib.mkEnableOption "userborn";
+
+    package = lib.mkOption {
+      type = lib.types.package;
+      default = pkgs.userborn;
+      defaultText = "pkgs.userborn";
+      description = "The userborn package";
+    };
+
+    immutablePasswordFilesLocation = lib.mkOption {
+      type = lib.types.str;
+      default = "/var/lib/nixos";
+      description = "The location of the original password files when using an immutable /etc";
+    };
+
+  };
+
+  config = lib.mkIf cfg.enable {
+
+    assertions = [
+      {
+        assertion = config.system.activationScripts.users == "";
+        message = "system.activationScripts.users has to be empty to use userborn";
+      }
+    ];
+
+    system.activationScripts.users = lib.mkForce "";
+    system.activationScripts.hashes = lib.mkForce "";
+
+    systemd = {
+
+      # Create home directories, do not create /var/empty even if that's a user's
+      # home.
+      tmpfiles.settings.home-directories = lib.mapAttrs'
+        (username: opts: lib.nameValuePair opts.home {
+          d = {
+            mode = opts.homeMode;
+            user = username;
+            inherit (opts) group;
+          };
+        })
+        (lib.filterAttrs (_username: opts: opts.home != "/var/empty") userCfg.users);
+
+      services.userborn = {
+        wantedBy = [ "sysinit.target" ];
+        requiredBy = [ "sysinit-reactivation.target" ];
+        after = [ "systemd-remount-fs.service" "systemd-tmpfiles-setup-dev-early.service" ];
+        before = [ "systemd-tmpfiles-setup-dev.service" "sysinit.target" "shutdown.target" "sysinit-reactivation.target" ];
+        conflicts = [ "shutdown.target" ];
+        restartTriggers = [ userbornConfigJson passwordFilesLocation ];
```

Comment:
I don't think the `passwordFilesLocation` bit here does anything? it's either `/etc/` or `/var/lib/nixos` neither which are nix store paths

Reply:
The only thing it does is that if your config doesn't change, but you switch from mutable to immutable /etc, the service get's restarted.